src/server/raftkv2/mod.rs

```diff
@@ -0,0 +1,301 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{mem, pin::Pin, task::Poll};
+
+use engine_traits::{KvEngine, RaftEngine, CF_LOCK};
+use futures::{Future, Stream, StreamExt};
+use kvproto::{
+    raft_cmdpb::{CmdType, RaftCmdRequest, Request},
+    raft_serverpb::RaftMessage,
+};
+use raftstore::store::RegionSnapshot;
+use raftstore_v2::{
+    router::{message::SimpleWrite, CmdResChannel, CmdResEvent, CmdResStream, PeerMsg, RaftRouter},
+    SimpleWriteEncoder, StoreRouter,
+};
+use tikv_kv::{Modify, RaftExtension, WriteEvent};
+use tikv_util::{codec::number::NumberEncoder, time::Instant};
+use txn_types::WriteBatchFlags;
+
+use super::{
+    metrics::{ASYNC_REQUESTS_COUNTER_VEC, ASYNC_REQUESTS_DURATIONS_VEC},
+    raftkv::{get_status_kind_from_engine_error, new_request_header},
+};
+
+#[derive(Clone)]
+pub struct RaftExtensionImpl<EK: KvEngine, ER: RaftEngine> {
+    router: StoreRouter<EK, ER>,
+}
+
+impl<EK: KvEngine, ER: RaftEngine> RaftExtension for RaftExtensionImpl<EK, ER> {
+    #[inline]
+    fn feed(&self, msg: RaftMessage, key_message: bool) {
+        let region_id = msg.get_region_id();
+        let msg_ty = msg.get_message().get_msg_type();
+        // Channel full and region not found are ignored unless it's a key message.
+        if let Err(e) = self.router.send_raft_message(Box::new(msg)) && key_message {
+            error!("failed to send raft message"; "region_id" => region_id, "msg_ty" => ?msg_ty, "err" => ?e);
+        }
+    }
+
+    fn report_reject_message(&self, _region_id: u64, _from_peer_id: u64) {
+        // TODO：reject the message on connection side instead of go through
+        // raft layer.
+    }
+
+    fn report_peer_unreachable(&self, region_id: u64, to_peer_id: u64) {
+        let _ = self
+            .router
+            .send(region_id, PeerMsg::PeerUnreachable { to_peer_id });
+    }
+
+    fn report_store_unreachable(&self, _store_id: u64) {}
+
+    fn report_snapshot_status(
+        &self,
+        _region_id: u64,
+        _to_peer_id: u64,
+        _status: raft::SnapshotStatus,
+    ) {
+    }
+
+    fn report_resolved(&self, _store_id: u64, _group_id: u64) {}
+
+    fn split(
+        &self,
+        _region_id: u64,
+        _region_epoch: kvproto::metapb::RegionEpoch,
+        _split_keys: Vec<Vec<u8>>,
+        _source: String,
+    ) -> futures::future::BoxFuture<'static, tikv_kv::Result<Vec<kvproto::metapb::Region>>> {
+        Box::pin(async move { Err(box_err!("raft split is not supported")) })
```

Comment:
I suppose split will be in another PR right? Maybe the error message could be "is not implemented yet."

Reply:
Yes, I just copied that from the definition of trait.